### PR TITLE
Specify max pods for lower instance types

### PR DESCRIPTION
**Pull request by njuettner**
When choosing smaller instances we want to make sure we schedule only
pods when we still have IP's available when using aws-cni.

Script:

```python
import requests
import json

from bs4 import BeautifulSoup

response = requests.get("https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-eni.html#AvailableIpPerENI")

parsed_html = BeautifulSoup(response.text, features="html.parser")

table = parsed_html.find('table', attrs={'id': 'w297aac21c13c15b5'})

rows = table.find_all("tr")

instance_max_pods = {}

for row in rows:
	cells = row.find_all("td")
	if len(cells) < 1:
		continue
	# (# of ENI - 1) * (Max IPs per ENI - 1) + 2
	instance_max_pods[cells[0].text.strip()] = (int(cells[1].text) - 1) * (int(cells[2].text) - 1) + 2
	# Add two for aws-node and kube-proxy hostNetwork pods
	# IPs per ENI - 1 = One IP address is allocated to the Host ENI itself.
print(json.dumps(instance_max_pods)
```

Amount of max pods when using aws-cni 

```json
{
  "a1.medium": 5,
  "a1.large": 20,
  "a1.xlarge": 44,
  "a1.2xlarge": 44,
  "a1.4xlarge": 205,
  "a1.metal": 205,
  "c1.medium": 7,
  "c1.xlarge": 44,
  "c3.large": 20,
  "c3.xlarge": 44,
  "c3.2xlarge": 44,
  "c3.4xlarge": 205,
  "c3.8xlarge": 205,
  "c4.large": 20,
  "c4.xlarge": 44,
  "c4.2xlarge": 44,
  "c4.4xlarge": 205,
  "c4.8xlarge": 205,
  "c5.large": 20,
  "c5.xlarge": 44,
  "c5.2xlarge": 44,
  "c5.4xlarge": 205,
  "c5.9xlarge": 205,
  "c5.12xlarge": 205,
  "c5.18xlarge": 688,
  "c5.24xlarge": 688,
  "c5.metal": 688,
  "c5a.large": 20,
  "c5a.xlarge": 44,
  "c5a.2xlarge": 44,
  "c5a.4xlarge": 205,
  "c5a.8xlarge": 205,
  "c5a.12xlarge": 205,
  "c5a.16xlarge": 688,
  "c5a.24xlarge": 688,
  "c5d.large": 20,
  "c5d.xlarge": 44,
  "c5d.2xlarge": 44,
  "c5d.4xlarge": 205,
  "c5d.9xlarge": 205,
  "c5d.12xlarge": 205,
  "c5d.18xlarge": 688,
  "c5d.24xlarge": 688,
  "c5d.metal": 688,
  "c5n.large": 20,
  "c5n.xlarge": 44,
  "c5n.2xlarge": 44,
  "c5n.4xlarge": 205,
  "c5n.9xlarge": 205,
  "c5n.18xlarge": 688,
  "c5n.metal": 688,
  "c6g.medium": 5,
  "c6g.large": 20,
  "c6g.xlarge": 44,
  "c6g.2xlarge": 44,
  "c6g.4xlarge": 205,
  "c6g.8xlarge": 205,
  "c6g.12xlarge": 205,
  "c6g.16xlarge": 688,
  "c6g.metal": 688,
  "cc2.8xlarge": 205,
  "cr1.8xlarge": 205,
  "d2.xlarge": 44,
  "d2.2xlarge": 44,
  "d2.4xlarge": 205,
  "d2.8xlarge": 205,
  "f1.2xlarge": 44,
  "f1.4xlarge": 205,
  "f1.16xlarge": 345,
  "g2.2xlarge": 44,
  "g2.8xlarge": 205,
  "g3s.xlarge": 44,
  "g3.4xlarge": 205,
  "g3.8xlarge": 205,
  "g3.16xlarge": 688,
  "g4dn.xlarge": 20,
  "g4dn.2xlarge": 20,
  "g4dn.4xlarge": 20,
  "g4dn.8xlarge": 44,
  "g4dn.12xlarge": 205,
  "g4dn.16xlarge": 44,
  "g4dn.metal": 688,
  "h1.2xlarge": 44,
  "h1.4xlarge": 205,
  "h1.8xlarge": 205,
  "h1.16xlarge": 688,
  "hs1.8xlarge": 205,
  "i2.xlarge": 44,
  "i2.2xlarge": 44,
  "i2.4xlarge": 205,
  "i2.8xlarge": 205,
  "i3.large": 20,
  "i3.xlarge": 44,
  "i3.2xlarge": 44,
  "i3.4xlarge": 205,
  "i3.8xlarge": 205,
  "i3.16xlarge": 688,
  "i3.metal": 688,
  "i3en.large": 20,
  "i3en.xlarge": 44,
  "i3en.2xlarge": 44,
  "i3en.3xlarge": 44,
  "i3en.6xlarge": 205,
  "i3en.12xlarge": 205,
  "i3en.24xlarge": 688,
  "i3en.metal": 688,
  "inf1.xlarge": 29,
  "inf1.2xlarge": 29,
  "inf1.6xlarge": 205,
  "inf1.24xlarge": 408,
  "m1.small": 5,
  "m1.medium": 7,
  "m1.large": 20,
  "m1.xlarge": 44,
  "m2.xlarge": 44,
  "m2.2xlarge": 89,
  "m2.4xlarge": 205,
  "m3.medium": 7,
  "m3.large": 20,
  "m3.xlarge": 44,
  "m3.2xlarge": 89,
  "m4.large": 11,
  "m4.xlarge": 44,
  "m4.2xlarge": 44,
  "m4.4xlarge": 205,
  "m4.10xlarge": 205,
  "m4.16xlarge": 205,
  "m5.large": 20,
  "m5.xlarge": 44,
  "m5.2xlarge": 44,
  "m5.4xlarge": 205,
  "m5.8xlarge": 205,
  "m5.12xlarge": 205,
  "m5.16xlarge": 688,
  "m5.24xlarge": 688,
  "m5.metal": 688,
  "m5a.large": 20,
  "m5a.xlarge": 44,
  "m5a.2xlarge": 44,
  "m5a.4xlarge": 205,
  "m5a.8xlarge": 205,
  "m5a.12xlarge": 205,
  "m5a.16xlarge": 688,
  "m5a.24xlarge": 688,
  "m5ad.large": 20,
  "m5ad.xlarge": 44,
  "m5ad.2xlarge": 44,
  "m5ad.4xlarge": 205,
  "m5ad.8xlarge": 205,
  "m5ad.12xlarge": 205,
  "m5ad.16xlarge": 688,
  "m5ad.24xlarge": 688,
  "m5d.large": 20,
  "m5d.xlarge": 44,
  "m5d.2xlarge": 44,
  "m5d.4xlarge": 205,
  "m5d.8xlarge": 205,
  "m5d.12xlarge": 205,
  "m5d.16xlarge": 688,
  "m5d.24xlarge": 688,
  "m5d.metal": 688,
  "m5dn.large": 20,
  "m5dn.xlarge": 44,
  "m5dn.2xlarge": 44,
  "m5dn.4xlarge": 205,
  "m5dn.8xlarge": 205,
  "m5dn.12xlarge": 205,
  "m5dn.16xlarge": 688,
  "m5dn.24xlarge": 688,
  "m5n.large": 20,
  "m5n.xlarge": 44,
  "m5n.2xlarge": 44,
  "m5n.4xlarge": 205,
  "m5n.8xlarge": 205,
  "m5n.12xlarge": 205,
  "m5n.16xlarge": 688,
  "m5n.24xlarge": 688,
  "m6g.medium": 5,
  "m6g.large": 20,
  "m6g.xlarge": 44,
  "m6g.2xlarge": 44,
  "m6g.4xlarge": 205,
  "m6g.8xlarge": 205,
  "m6g.12xlarge": 205,
  "m6g.16xlarge": 688,
  "m6g.metal": 688,
  "p2.xlarge": 44,
  "p2.8xlarge": 205,
  "p2.16xlarge": 205,
  "p3.2xlarge": 44,
  "p3.8xlarge": 205,
  "p3.16xlarge": 205,
  "p3dn.24xlarge": 688,
  "r3.large": 20,
  "r3.xlarge": 44,
  "r3.2xlarge": 44,
  "r3.4xlarge": 205,
  "r3.8xlarge": 205,
  "r4.large": 20,
  "r4.xlarge": 44,
  "r4.2xlarge": 44,
  "r4.4xlarge": 205,
  "r4.8xlarge": 205,
  "r4.16xlarge": 688,
  "r5.large": 20,
  "r5.xlarge": 44,
  "r5.2xlarge": 44,
  "r5.4xlarge": 205,
  "r5.8xlarge": 205,
  "r5.12xlarge": 205,
  "r5.16xlarge": 688,
  "r5.24xlarge": 688,
  "r5.metal": 688,
  "r5a.large": 20,
  "r5a.xlarge": 44,
  "r5a.2xlarge": 44,
  "r5a.4xlarge": 205,
  "r5a.8xlarge": 205,
  "r5a.12xlarge": 205,
  "r5a.16xlarge": 688,
  "r5a.24xlarge": 688,
  "r5ad.large": 20,
  "r5ad.xlarge": 44,
  "r5ad.2xlarge": 44,
  "r5ad.4xlarge": 205,
  "r5ad.8xlarge": 205,
  "r5ad.12xlarge": 205,
  "r5ad.16xlarge": 688,
  "r5ad.24xlarge": 688,
  "r5d.large": 20,
  "r5d.xlarge": 44,
  "r5d.2xlarge": 44,
  "r5d.4xlarge": 205,
  "r5d.8xlarge": 205,
  "r5d.12xlarge": 205,
  "r5d.16xlarge": 688,
  "r5d.24xlarge": 688,
  "r5d.metal": 688,
  "r5dn.large": 20,
  "r5dn.xlarge": 44,
  "r5dn.2xlarge": 44,
  "r5dn.4xlarge": 205,
  "r5dn.8xlarge": 205,
  "r5dn.12xlarge": 205,
  "r5dn.16xlarge": 688,
  "r5dn.24xlarge": 688,
  "r5n.large": 20,
  "r5n.xlarge": 44,
  "r5n.2xlarge": 44,
  "r5n.4xlarge": 205,
  "r5n.8xlarge": 205,
  "r5n.12xlarge": 205,
  "r5n.16xlarge": 688,
  "r5n.24xlarge": 688,
  "r6g.medium": 5,
  "r6g.large": 20,
  "r6g.xlarge": 44,
  "r6g.2xlarge": 44,
  "r6g.4xlarge": 205,
  "r6g.8xlarge": 205,
  "r6g.12xlarge": 205,
  "r6g.16xlarge": 688,
  "r6g.metal": 688,
  "t1.micro": 3,
  "t2.nano": 3,
  "t2.micro": 3,
  "t2.small": 8,
  "t2.medium": 12,
  "t2.large": 24,
  "t2.xlarge": 30,
  "t2.2xlarge": 30,
  "t3.nano": 3,
  "t3.micro": 3,
  "t3.small": 8,
  "t3.medium": 12,
  "t3.large": 24,
  "t3.xlarge": 44,
  "t3.2xlarge": 44,
  "t3a.nano": 3,
  "t3a.micro": 3,
  "t3a.small": 5,
  "t3a.medium": 12,
  "t3a.large": 24,
  "t3a.xlarge": 44,
  "t3a.2xlarge": 44,
  "u-6tb1.metal": 118,
  "u-9tb1.metal": 118,
  "u-12tb1.metal": 118,
  "u-18tb1.metal": 688,
  "u-24tb1.metal": 688,
  "x1.16xlarge": 205,
  "x1.32xlarge": 205,
  "x1e.xlarge": 20,
  "x1e.2xlarge": 44,
  "x1e.4xlarge": 44,
  "x1e.8xlarge": 44,
  "x1e.16xlarge": 205,
  "x1e.32xlarge": 205,
  "z1d.large": 20,
  "z1d.xlarge": 44,
  "z1d.2xlarge": 44,
  "z1d.3xlarge": 205,
  "z1d.6xlarge": 205,
  "z1d.12xlarge": 688,
  "z1d.metal": 688
}
```

## Checklist

- [ ] Update changelog in CHANGELOG.md.

### CODE REVIEW

**Comment by kopiczko** [comment]
This looks similar to https://github.com/giantswarm/giantnetes-terraform/pull/331

Not sure if we can somehow align that.

**Comment by njuettner** [comment]
I didn't know about that. I thought this is the source of truth to modify max pods. I'm fine closing this PR as long as we make sure we don't set max pods too high when aws-cni  is enabled

**Comment by kopiczko** [comment]
> I thought this is the source of truth to modify max pods.

The thing I liked is for CP. This is for TC so this PR is valid.

**Comment by njuettner** [comment]
More or less the same as we use for CP cluster except we don't want to have more than 110 pods on a node. Talked with @calvix about that and this should be fine.

**Comment by marians** [comment]
The HTML parsing part looks a bit fragile. How about using the JSON from https://ec2instances.info/?min_memory=2&min_vcpus=2&region=eu-central-1 ? Raw data in https://github.com/powdahound/ec2instances.info/blob/master/www/instances.json

File: files/conf/setup-kubelet-environment [hunk 1]
```diff
@@ -8,15 +8,311 @@ EnvFile="/etc/kubelet-environment"
...
-instance_size=$(curl http://169.254.169.254/latest/meta-data/instance-type 2>/dev/null| cut -d\. -f2)
+instance_type="$(curl http://169.254.169.254/latest/meta-data/instance-type 2>/dev/null)"
 
-if [[ "${instance_size}" =~ ^(4xlarge|8xlarge|9xlarge|12xlarge|16xlarge|18xlarge|24xlarge)$ ]]; then
-	MAX_PODS=110
-else 
-	MAX_PODS=40
+# Cases are generated with:
+#
+#    curl -fsSLo - https://raw.githubusercontent.com/awslabs/amazon-eks-ami/2e1f63f951c82a76fd20b19b811592535962c82d/files/eni-max-pods.txt | grep -v '^#' | sed -E 's/([[:alnum:]]+)[[:space:]]+([[:digit:]]+)/\1)|MAX_PODS=$((\2|- $reserved_ips)) ;;/' | column -t -s '|'
```
[the remark this answers is not in the document]
Comment (njuettner) [reply files/conf/setup-kubelet-environment:15]: Good idea we should iterate on that and make it a bit easier.

**Comment by njuettner** [comment]
> The HTML parsing part looks a bit fragile. How about using the JSON from https://ec2instances.info/?min_memory=2&min_vcpus=2&region=eu-central-1 ? Raw data in https://github.com/powdahound/ec2instances.info/blob/master/www/instances.json

I agree with your point, it's worth choosing another way. This was mainly to demonstrate the available ENI's we can get from each instance. 

The code itself in the PR took the same settings as from the tenant cluster (giantswarm/giantnetes-terraform#331).

**Comment by njuettner** [comment]
I'll merge it for now and check the idea from @marians by using ec2instances.info.

**Pull request merged**